### PR TITLE
The inferior limit (lim inf)

### PR DESCRIPTION
**Pull request by glacode**
among the new theorems, some that could be of interest:

- "basic" properties: ~ liminfcl , ~ liminflelimsupuz

- relationships between convergence, limsup and liminf: from ~ climliminflimsup to ~ climliminflimsup4

- the inferior limit of a countable set of sigma-measurable functions is sigma-measurable: ~ smfliminf

### CODE REVIEW

**Comment by glacode** [comment]
Mathbox only (I feel so stupid I cannot find a way to add the nice green label...)

**Comment by benjub** [comment]
> Mathbox only (I feel so stupid I cannot find a way to add the nice green label...)

It's a cranked wheel next to "Labels" on the right panel, but maybe permissions are needed for it to be available.

**Comment by avekens** [comment]
> Mathbox only (I feel so stupid I cannot find a way to add the nice green label...)

I think you need special permissions for setting labels. I also cannot do it ...

**Comment by benjub** [comment]
@glacode Do you want to modify ~climlimsupcex in view of the discussion group thread ?

There are about 2100 lines to review (cursorily, since it's in a mathbox). Can we split ? @avekens ?  @tirix ?  700 lines each ?

**Review by avekens — APPROVED** [review]
I had a look at the changes up to line 658261 (no remarks), but stopped at 0cnv because of the google group discussion. Therefore, I give my approval up to line 658261

**Comment by benjub** [comment]
> I had a look at the changes up to line 658261 (no remarks), but stopped at 0cnv because of the google group discussion. Therefore, I give my approval up to line 658261

Thanks. I call dibs on 658261--663996.

File: set.mm
Comment (benjub) [inline set.mm]: See discussion group thread. In set.mm, one has `|- -. (/) e. CC`, therefore one has `|- ( (/) e. CC -> ph )`. If your proof exhibits a special feature, then maybe extract that subproof to state precisely that.

File: set.mm
Comment (benjub) [inline set.mm]: Maybe label it climuzv and add to the comment a sentence: "This is the same as ~ climuz with a dv condition on ` k , F ` instead of a non-freeness hypothesis." [if I understood correctly]

File: set.mm
Comment (benjub) [inline set.mm:663512]: superior --> inferior
The comment is a bit strange (I see it is the same for limsupgord by @digama0). This statement is not about inferior/superior limit: it is simply used in some liminf/limsup results, but it merely says that the sup on a larger set is larger. What about "The infimum of an extended real function is a non-increasing function from uppersets of reals ordered by inclusion to the extended reals." and similarly with limsupgord ? @digama0 ?

It would be nice (future project) to prove these from the general theorem:
(A and B are subsets of an ordered set (R,<) with A C_ B) -> sup (A,R,<) <= sup (B,R,<) [defining <= to be "< or equal"]
and deduce from it
(f: X --> R and (R, <) is an ordered set and A and B are subsets of X with A C_ B) -> sup (F"A,R,<) <= sup (F"B,R,<)
and then our theorems.

More generally: I haven't checked the details of your proofs, but I imagine you took advantage of the fact that liminf for the standard order is limsup for the opposite order, to reuse the theorems for limsup ?

File: set.mm
Comment (benjub) [inline set.mm:663485]: an infinite sequence --> a sequence ?
(actually, it makes sense for any extended real function defined on a subset of RR which is not upper-bounded) [<-- you can add this to the comment if you want]

File: set.mm
Comment (benjub) [inline set.mm:663531]: an infinite sequence --> a sequence (or as in the comment of df-liminf)
, which is --> is
... of all subsequences of F defined on an upperset (or, if you consider any function F as in the comment of df-liminf, "... of all restrictions of F to an upperset of reals")

File: set.mm
Comment (benjub) [inline set.mm:663545]: restricted to an upperset of reals.

File: set.mm
Comment (benjub) [inline set.mm:663556]: A superior limit is an extended real.

File: set.mm
Comment (benjub) [inline set.mm:663566]: I don't understand why "closure" is used here, but I see it is in many other places in set.mm so maybe it comes from computer science. Leave it if it makes sense to you. Also, "inferior limit function" may not be the best description, but leave it if it's consistent with your sources.

File: set.mm
Comment (benjub) [inline set.mm:663593]: I would remove "The first..." (see discussion group), since when writing ( ZZ>= ` M ) , it is more or less expected that M e. ZZ (with the current definition; one could adopt a definition of ZZ>= making sense for any argument in RR*).

File: set.mm
Comment (benjub) [inline set.mm:663680]: ...only depends on the restriction of that function to the preimage of the set of extended reals.
(same thing in the next statement)

File: set.mm
Comment (benjub) [inline set.mm:663769]: Again, see if you want to change this in view of the discussion group thread.

File: set.mm
Comment (benjub) [inline set.mm:663795]: Same comment as above.

File: set.mm
Comment (benjub) [inline set.mm:663986]: appxoximated --> approximated (same above)
I would write something like: "For any positive real, the superior limit of F is larger than any of its values at large enough arguments, up to that positive real."

**Review by benjub — COMMENTED** [review]
Reviewed lines 658261--663996, and my comments are only suggestions, since it's in your mathbox.
Also, when a statement is simply a lemma labeled e.g. xxxlem, the comment may simply say "Lemma for xxx", possibly with a comment about the lemma, but do not repeat the comment of xxx in xxxlem.  I noticed that many of your lemmas can be commented as: "Lemma for xxx : same conclusion as xxx but with dv conditions on... / with the hypothesis that..."

File: set.mm
Comment (glacode) [reply set.mm]: > See discussion group thread.

Which thread? Can you please post a link?

set.mm has

|- -. (/) e. CC

but it is a construction-dependent property, in fact new usage is discouraged.

Here is Norman's comment in ~0ncn

"Note: do not use this after the real number axioms are developed, since it is a construction-dependent property."

Thus, actually we don't have the theorem

|- -. (/) e. CC

On the contrary,

|- ( (/) e. CC -> (/) ~~> (/) )

would be proven with any arbitrary construction of the complex numbers (it is construction independent). That's why I put it in.

File: set.mm
Comment (glacode) [reply set.mm:663566]: I guess, I simply adopted Mario's comment for limsupgf (of course, I changed superior with inferior)

File: set.mm
Comment (benjub) [reply set.mm]: This one: https://groups.google.com/g/metamath/c/BXJq5o5i8a8/m/WowODKouCgAJ

After looking at df-clim, I think I can see what you're doing (I haven't looked at your proof yet, I'll do it when it's merged).  In the expression (/) ~~> (/), the (/) on the left is the empty function, the (/) on the right is a complex number, and you're probably using the fact that a function evaluates outside its domain to the empty set (which is due to the particular encoding used in set.mm), so this is abusing the encoding used in set.mm. It's not meant to be a criticism at all, but I think we agree there is no mathematical content in this theorem ?  Of course feel free to leave it as is in your mathbox.  If you think I'm missing something, I'll look at the proof once on the website.

File: set.mm
Comment (glacode) [reply set.mm:663531]: I believe I have used the same comment as ~ limsupval (for consistency)

To say the truth, I was not so happy about neither the "sequence" nor the "subsequence" terms, because the definition would be "meaningful" for any set: in particular, for functions with dense domain in RR (or defined on all RR).
But the "standard" use will be with sequences, that's why I adapted Norman's comment.

File: set.mm
Comment (glacode) [reply set.mm:663545]: Much better, thank you for pointing it out. I'll fix it.

File: set.mm
Comment (glacode) [reply set.mm:663556]: I simply adopted Norman's comment for ~ limsupcl  (for consistency)

I agree your comment provides more information (but may be "Closure of..." is a "standard" comment in set.mm ? )

File: set.mm
Comment (glacode) [reply set.mm]: > This one: https://groups.google.com/g/metamath/c/BXJq5o5i8a8/m/WowODKouCgAJ
> 
> In the expression (/) ~~> (/), the (/) on the left is the empty function, the (/) on the right is a complex number

Exactly. I was investigating the behavior of ~~> when applied to the empty function.

Thinking about it, I guess I could complete it with another theorem like

|- ( -. (/) e. CC -> -.  (/) e. dom ~~> )

that is, "assuming that the empty set is not a complex number, the empty function doesn't converge".

(but I've not proven it yet, so I'm not sure about it).

I absolutely agree, there's no mathematical content in this theorem, just a couple of assertions to clarify the behavior of our particular definition of convergence when applied to the empty set.
Of course, our definition of convergence agrees with what we find in textbooks for the "standard" case, i.e. sequences, but with metamath we often have to deal with edge cases that textbooks overlook, so it might be useful to write it once and keep if for future reference: if, in the future, someone will have to deal with an edge case and will search for "(/) ~~>" , she will see that it depends on an "undecidable" antecedent and look no further.

Comment (glacode) [reply set.mm]: w.r.t. the thread, I asked about

-. +oo e. CC

because, while for -. (/) e. CC it is clear to me Norman's intent to state it as something that we CANNOT rely upon (no textbook will ever "take a position" about the empty set being a complex number or not), it was not so clear to me if math literature had an "official" position about chosing +oo and -oo distinct not only from the reals, but from the complex numbers too.

Comment (benjub) [reply set.mm]: Since we don't want the empty function to converge to anything, your theorem shows that we'd better require that -. (/) e. CC, as is rightly the case (or that the value of a function out of its domain be not the empty set). Similarly, we'd better have +oo not in any other set that could come up, to avoid clashes and "accidental" or "junk" theorems (for instance: "plus infinity belongs to the algebraic closure of RR"!!). See for instance https://groups.google.com/g/metamath/c/GQmelsKtaWQ

**Comment by avekens** [comment]
There was a discussion about the term "closure" in this PR (I cannot find it anymore - is it removed, because it is resolved?).

I think the term "closure" is a little misused in set.mm, many times meaning simply "membership". But it in the original sense, it means "if the argument/operands belong to a subset of the domain the function/operation value also belongs to this subset (see Wikipdia "Closure (mathematics)": "a subset of a given set is **closed** under an operation of the larger set if performing that operation on members of the subset always produces a member of that subset. . See, for example ~zaddcl: ( ( M e. ZZ /\ N e. ZZ ) -> ( M + N ) e. ZZ ) with + : ( CC X. CC ) --> CC. 

For ~ceilcl, for example, I think the usage of "closure" is maybe not correct: ( A e. RR -> ( |^ ` A ) e. ZZ ) This theorem simply states that the function value belongs to the range/codomain of the function. In this case, however, there is no theorem |^ : RR --> ZZ, but it could be simply derived from ~ceilcl. Therefore, I think the term "closure" is exactly used also for such cases: "The range of a function/operation is (a subset of) a special class".

**Comment by benjub** [comment]
@avekens wrt "closure": you're right about the correct usage of "closure" (first paragraph). Other example: "a topology is a family of subsets *closed under* unions and finite intersections".  There is also a use in set.mm of "reverse closure" e.g. things like (x+y) e. z -> x e. z.

I agree that "closure" is used a bit too widely in set.mm, and ~ceicl, the usage of "closure" is not correct.  This can be corrected on a case by case basis if you encounter them.

File: set.mm
Comment (benjub) [inline set.mm:664055]: Function --> function

File: set.mm
Comment (benjub) [inline set.mm:664136]: function

File: set.mm
Comment (glacode) [reply set.mm:663512]: I've fixed superior --> inferior (see new commit)
If Mario agrees, we can change both comments as you suggested (I'd prefert to do it after the merge, so that I can dedicate a new pull request just to clean up the mathbox and minimize the risk of conflicts, becasue this pending PR is pretty large)

File: set.mm
Comment (glacode) [reply set.mm:663485]: In my last commit:
- an infinite sequence --> a sequence
- added (actually, it makes sense for any extended real function defined on a subset of RR which is not upper-bounded)
I like the additional comment, I wasn't so happy with "sequence" because the definition is more general than that.

File: set.mm
Comment (glacode) [reply set.mm:663531]: last commit:
- I've replaced an infinite sequence --> a sequence
- , which is --> is
- now it is    ... of the set of suprema of all restrictions of ` F ` to an upperset of reals

File: set.mm
Comment (glacode) [reply set.mm:663680]: Changed as suggested (in the last commit)

File: set.mm
Comment (glacode) [reply set.mm:663795]: New commit, now it is:
The inferior limit doesn't change when a function is restricted to an upperset of reals.

File: set.mm
Comment (glacode) [reply set.mm:663986]: New commit:
changed as suggested

File: set.mm
Comment (benjub) [inline set.mm:664228]: when the given function

**Review by benjub — APPROVED** [review]
Approved after a very cursory look at the remaining third of the PR.  I made a few suggestions related to comments.  If you have time and haven't done so, you may do a "MM>minimize_with *" on these proofs.

File: set.mm
Comment (glacode) [reply set.mm:664055]: fixed in e5878437

File: set.mm
Comment (glacode) [reply set.mm:664136]: fixed in e5878437

File: set.mm
Comment (glacode) [reply set.mm:664228]: fixed in e5878437 (thank you for the revision!)

**Pull request merged**
